Feat(Observations): Add automatic translations

santostiago opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help -->

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [ ] Check the commit's or even all commits' message styles matches our requested structure.
- [ ] Check your code additions will fail neither code linting checks nor unit test.



## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

santostiago commented:
@tsubik I'll add the specs during the week, but the code should be close to what it is now.

tsubik reviewed (commented):
Cannot run `bundle install`. There is some problem with `googleauth (1.9.0)`, not sure why. Quick fix is to just run `bundle update google-cloud-translate`.

tsubik commented on app/jobs/translation_job.rb:
```diff
@@ -0,0 +1,40 @@
+      I18n.available_locales.each do |locale|
+        next if locale == original_locale
+
+        translated_fields[field][locale] = translation_service.run(entity.send(field), original_locale, locale)
```
Only this line could be within block of `I18n.with_locale(original_locale)`. Instead of changing locale globally couple lines above. I think that could be more clear what is going on here.

tsubik commented on db/migrate/20240128163512_add_automatic_translation_field_to_observation_translations.rb:
```diff
@@ -0,0 +1,7 @@
+class AddAutomaticTranslationFieldToObservationTranslations < ActiveRecord::Migration[7.0]
+  def change
+    add_column :observation_translations, :details_translated_from, :string
```
All columns are translated at once, couldn't we have only one column like `auto_translated_from`?

santostiago replied:
I thought of it, but maybe we'd like to change this to translate just one attribute, and it would be nice to know which one.

tsubik replied:
Sure, I thought that could be the reason.

tsubik reviewed (commented):
Looks really good. I still need to test it locally after setting up google cloud. For now, I'm adding a couple comments. Also, this needs rebasing (that would fix security workflow).

tsubik commented on app/admin/observation.rb:
```diff
@@ -581,10 +586,16 @@ def permitted_params
 
     f.inputs I18n.t("active_admin.shared.translated_fields") do
+      div do
+        link_to I18n.t("active_admin.observations_page.force_translations"), force_translations_admin_observation_path(resource), method: :put, class: 'button'
```
There is also problem that user using English in backend interface, but is changing something in French and wants to force translation from French not English.

santostiago replied:
Mmmm.. from the conversation with Sophie I understood she'd like to have a button to make the translation, instead of saving the whole form.

santostiago commented on app/resources/v1/observation_resource.rb:
```diff
@@ -120,10 +120,11 @@ def set_user
     def set_modified
       user = context[:current_user]
       @model.modified_user_id = user.id
+      @model.admin_locale = user.locale
```
Right!

santostiago replied:
In fact, we have a lot of observations without `modfied_user`.

tsubik replied:
Maybe those are old ones or were imported at the beginning?

santostiago commented on app/jobs/translation_job.rb:
```diff
@@ -0,0 +1,40 @@
+
+    translated_fields.each do |field, translation|
+      translation.each do |locale, text|
+        I18n.locale = locale
```
You mean the block? Yeah, can be.

tsubik replied:
Yes, the block.

tsubik commented on .env.sample:
```diff
@@ -23,3 +23,6 @@ RESPONSIBLE_EMAIL=
 
 MAPBOX_API_KEY=
+
+GOOGLE_CLIENT_ID=
```
looks like only API_KEY is used, so this could be removed.

santostiago commented:
@tsubik I added a new field on the edit form containing the language to translate from. When something is selected, when submitting the form, the translation will occur based on the selected language.
I also added a way to directly translate the field on the view page.

santostiago commented:
It seems like all the tests are failing... maybe I should rebase from master.

tsubik commented on app/admin/observation.rb:
```diff
@@ -581,10 +597,20 @@ def permitted_params
     end
 
     f.inputs I18n.t("active_admin.shared.translated_fields") do
+      div do
```
The input looks weird, but if you remove that `div` it looks a way better, not sure why that div is here?

tsubik commented on app/admin/observation.rb:
```diff
@@ -581,10 +597,20 @@ def permitted_params
 
     f.inputs I18n.t("active_admin.shared.translated_fields") do
+      div do
+        f.input :admin_locale, label:  I18n.t("active_admin.observations_page.translate_from"),
```
I like it that it's only select input without the button to force translation.

tsubik replied:
I would hide it when observation is not `published`, as I see in that case force translation is not working. I would also add some hint message saying something about how it works.

tsubik replied:
Hint message like: "Set if you want to perform automatic translations of all texts from selected language"

tsubik reviewed (changes requested):
I like the new way of forcing translations, but I have some errors with translating job. Not sure what is the problem (some invalid argument error coming from google translate), will check it again tomorrow (maybe something changed in gem version as I have to `bundle update google-cloud-translate` to even run this code). I left some other comments though.

tsubik commented on app/jobs/translation_job.rb:
```diff
@@ -0,0 +1,43 @@
+    fields.each do |field|
+      translated_fields[field] = {}
+      I18n.available_locales.each do |locale|
+        next if locale == original_locale
```
Ok, I found what is the problem. It still tries to do translations like from "fr" to "fr" as when forcing translations `original_locale` is string and available_locales symbol, needs to be normalized.

tsubik commented on app/models/observation.rb:
```diff
@@ -85,6 +87,7 @@ class Observation < ApplicationRecord
   ].freeze
 
   attr_accessor :user_type
+  attr_accessor :admin_locale
```
I would probably change that name to `force_translations_from` or similar.

tsubik commented on app/admin/observation.rb:
```diff
@@ -103,6 +103,20 @@ def permitted_params
   end
 
+  member_action :force_translations do
+    translate_to = params[:translate_to] || I18n.locale
```
Hmm, shouldn't that be `translate_from`?

tsubik commented on app/admin/observation.rb:
```diff
@@ -103,6 +103,20 @@ def permitted_params
+  end
+
+  action_item :force_translations, only: :show do
+    dropdown_menu I18n.t("active_admin.observations_page.force_translations") do
```
Nice.

tsubik commented on app/admin/observation.rb:
```diff
@@ -103,6 +103,20 @@ def permitted_params
+  action_item :force_translations, only: :show do
+    dropdown_menu I18n.t("active_admin.observations_page.force_translations") do
+      I18n.available_locales.each do |locale|
+        item locale, force_translations_admin_observation_path(observation, translate_to: locale)
```
same, should probably be `translate_from`.

tsubik reviewed (commented):
Just found some minor things, probably naming typos. Otherwise, force translations feature works great.

tsubik reviewed (approved):
LGTM :+1:
